firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java

```diff
@@ -0,0 +1,169 @@
+        new BloomFilter(new byte[] {(byte) 255, (byte) 255, (byte) 255}, 1, 16);
+
+    assertFalse(emptyBloomFilter.mightContain(""));
+    assertFalse(nonEmptyBloomFilter.mightContain(""));
+  }
+
+  /**
+   * Golden tests are generated by backend based on inserting n number of document paths into a
+   * bloom filter.
+   *
+   * <p>Full document path is generated by concatenating documentPrefix and number n, eg,
+   * projects/project-1/databases/database-1/documents/coll/doc12.
+   *
+   * <p>The test result is generated by checking the membership of documents from documentPrefix+0
+   * to documentPrefix+2n. The membership results from 0 to n is expected to be true, and the
+   * membership results from n to 2n is expected to be false with some false positive results.
+   */
+  @Test
+  @SuppressWarnings("DefaultCharset")
+  public void testBloomFilterGoldenTest() throws Exception {
+    String documentPrefix = "projects/project-1/databases/database-1/documents/coll/doc";
+
+    // Import the golden test files for bloom filter
+    HashMap<String, JSONObject> parsedSpecFiles = new HashMap<>();
+    File jsonDir = new File("src/test/resources/bloom_filter_golden_test_data");
+    File[] jsonFiles = jsonDir.listFiles();
+    for (File f : jsonFiles) {
+      if (!f.toString().endsWith(".json")) {
+        continue;
+      }
+
+      // Read the files into a map.
+      StringBuilder builder = new StringBuilder();
+      BufferedReader reader = new BufferedReader(new FileReader(f));
+      Stream<String> lines = reader.lines();
+      lines.forEach(builder::append);
+      String json = builder.toString();
+      JSONObject fileJSON = new JSONObject(json);
+      parsedSpecFiles.put(f.getName(), fileJSON);
+    }
+
+    // Loop and test the files
+    for (String fileName : parsedSpecFiles.keySet()) {
+      if (fileName.contains("membership_test_result")) {
+        continue;
+      }
+
+      // Read test data and instantiate a BloomFilter object
+      JSONObject fileJSON = parsedSpecFiles.get(fileName);
+      JSONObject bits = fileJSON.getJSONObject("bits");
+      String bitmap = bits.getString("bitmap");
+      int padding = bits.getInt("padding");
+      int hashCount = fileJSON.getInt("hashCount");
+      BloomFilter bloomFilter =
+          new BloomFilter(Base64.getDecoder().decode(bitmap), padding, hashCount);
+
+      // Find corresponding membership test result.
+      JSONObject resultJSON =
+          parsedSpecFiles.get(fileName.replace("bloom_filter_proto", "membership_test_result"));
+      String membershipTestResults = resultJSON.getString("membershipTestResults");
+
+      // Run and compare mightContain result with the expectation.
+      for (int i = 0; i < membershipTestResults.length(); i++) {
+        boolean expectedMembershipResult = membershipTestResults.charAt(i) == '1';
+        boolean mightContain = bloomFilter.mightContain(documentPrefix + i);
+        assertEquals(mightContain, expectedMembershipResult);
```

Comment:
Use Truth here instead of the base assertEquals. Using Truth I believe there is a way to attach some contextual information to the failure message. It would be really helpful if the failure message included the name of the golden file and the value of `i` that failed.

Reply:
AssertEquals can take messages as well, and java keeps suggesting me simplify assertTrue with assertEquals.